File: analysis/R/shrinkage.R
# M7 (Eager & Erickson, Ch.9) — multilevel / mixed-effects shrinkage in R.
# M1 showed rushing efficiency is noisy; M7 fixes what to DO about it. A random-intercept model
# on per-play RYOE partially pools each rusher toward the league mean; the pull is largest for
# small-sample rushers (regression to the mean). The BLUP is the rusher's shrunken RYOE — the
# estimate you'd actually trust over the raw average.
#   lmer(ryoe ~ 1 + (1 | rusher))
#
# Reads:  data/gold/plays_model.csv, data/results/pred__ryoe__r.csv
# Writes: data/results/pred__shrinkage__r.csv, coef__shrinkage__r.csv, metrics__shrinkage__r.csv

source("analysis/R/util_io.R")
suppressPackageStartupMessages({ library(dplyr); library(lme4) })
set.seed(42)

MIN_CARRIES <- 5

plays <- cfb_read_gold("plays_model") |>
  filter(is_rush == 1, !is.na(rusher_player_name)) |>
  select(play_key, rusher = rusher_player_name)
ryoe <- readr::read_csv(file.path("data", "results", "pred__ryoe__r.csv"),
                        show_col_types = FALSE) |>
  select(play_key, ryoe)

d <- inner_join(ryoe, plays, by = "play_key") |>
  group_by(rusher) |> filter(n() >= MIN_CARRIES) |> ungroup()

m <- lmer(ryoe ~ 1 + (1 | rusher), data = d)

grand <- as.numeric(fixef(m)[1])
re <- ranef(m)$rusher                        # per-rusher intercept deviation (shrunken)
vc <- as.data.frame(VarCorr(m))
tau <- vc$sdcor[vc$grp == "rusher"]          # between-rusher SD
sigma <- vc$sdcor[vc$grp == "Residual"]      # within-rusher (play) SD

raw <- d |> group_by(rusher) |>
  summarise(carries = n(), raw_ryoe = mean(ryoe), .groups = "drop")
shrunk <- tibble(rusher = rownames(re), shrunk_ryoe = grand + re[, 1])

pred <- raw |>
  inner_join(shrunk, by = "rusher") |>
  mutate(shrinkage = raw_ryoe - shrunk_ryoe, language = "r")

coef <- data.frame(model = "shrinkage", term = "(Intercept)", estimate = grand, language = "r")
metrics <- data.frame(
  model = "shrinkage",
  metric = c("grand_mean_ryoe", "tau_between_rusher_sd", "sigma_residual_sd",
             "icc", "n_rushers", "n_plays"),
  value = c(grand, tau, sigma, tau^2 / (tau^2 + sigma^2), nrow(pred), nrow(d)),
  language = "r")

cfb_write_result("pred__shrinkage__r", pred)
cfb_write_result("coef__shrinkage__r", coef)
cfb_write_result("metrics__shrinkage__r", metrics)
cat(sprintf("[shrinkage/R] tau=%.3f sigma=%.3f ICC=%.4f; %d rushers (>=%d carries)\n",
            tau, sigma, tau^2 / (tau^2 + sigma^2), nrow(pred), MIN_CARRIES))

File: analysis/python/shrinkage.py
"""M7 (Eager & Erickson, Ch.9) — multilevel shrinkage, Python parity of shrinkage.R.

statsmodels MixedLM fits the same random-intercept model as lme4 (both REML by default), so the
fixed intercept and variance components should agree closely and the per-rusher BLUPs correlate
~1 with R. Mixed-model optimizers differ, so this is reported as agreement metrics, not the
strict coefficient gate.

Reads data/gold/plays_model.csv, data/results/pred__ryoe__py.csv (+ the R output for the parity
report); writes pred__shrinkage__py / coef__shrinkage__py / metrics__shrinkage__py /
metrics__shrinkage_parity__py.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf

from cfb_analytics.config import REPO_ROOT

GOLD = REPO_ROOT / "data" / "gold"
RESULTS = REPO_ROOT / "data" / "results"
MIN_CARRIES = 5


def main() -> None:
    plays = pd.read_csv(GOLD / "plays_model.csv")
    plays = plays[(plays.is_rush == 1) & plays.rusher_player_name.notna()][
        ["play_key", "rusher_player_name"]].rename(columns={"rusher_player_name": "rusher"})
    ryoe = pd.read_csv(RESULTS / "pred__ryoe__py.csv")[["play_key", "ryoe"]]

    d = ryoe.merge(plays, on="play_key")
    d = d.groupby("rusher").filter(lambda g: len(g) >= MIN_CARRIES)

    md = smf.mixedlm("ryoe ~ 1", d, groups=d["rusher"]).fit()
    grand = float(md.fe_params["Intercept"])
    tau = float(np.sqrt(md.cov_re.iloc[0, 0]))     # between-rusher SD
    sigma = float(np.sqrt(md.scale))               # residual (play) SD

    re = {g: float(v.iloc[0]) for g, v in md.random_effects.items()}
    raw = d.groupby("rusher")["ryoe"].agg(carries="count", raw_ryoe="mean").reset_index()
    raw["shrunk_ryoe"] = raw["rusher"].map(lambda g: grand + re.get(g, 0.0))
    raw["shrinkage"] = raw["raw_ryoe"] - raw["shrunk_ryoe"]
    raw["language"] = "py"

    coef = pd.DataFrame([dict(model="shrinkage", term="(Intercept)",
                              estimate=grand, language="py")])
    icc = tau**2 / (tau**2 + sigma**2)
    metrics = pd.DataFrame([
        dict(model="shrinkage", metric="grand_mean_ryoe", value=grand),
        dict(model="shrinkage", metric="tau_between_rusher_sd", value=tau),
        dict(model="shrinkage", metric="sigma_residual_sd", value=sigma),
        dict(model="shrinkage", metric="icc", value=icc),
        dict(model="shrinkage", metric="n_rushers", value=float(len(raw))),
        dict(model="shrinkage", metric="n_plays", value=float(len(d))),
    ])
    metrics["language"] = "py"

    RESULTS.mkdir(parents=True, exist_ok=True)
    raw.to_csv(RESULTS / "pred__shrinkage__py.csv", index=False)
    coef.to_csv(RESULTS / "coef__shrinkage__py.csv", index=False)
    metrics.to_csv(RESULTS / "metrics__shrinkage__py.csv", index=False)

    # agreement vs R: fixed intercept, variance components, BLUP correlation
    r_path = RESULTS / "pred__shrinkage__r.csv"
    if r_path.exists():
        r = pd.read_csv(r_path)[["rusher", "shrunk_ryoe"]].rename(
            columns={"shrunk_ryoe": "shrunk_r"})
        merged = raw.merge(r, on="rusher")
        blup_corr = float(np.corrcoef(merged["shrunk_ryoe"], merged["shrunk_r"])[0, 1])
        rm = pd.read_csv(RESULTS / "metrics__shrinkage__r.csv").set_index("metric")["value"]
        parity = pd.DataFrame([
            dict(model="shrinkage", metric="blup_corr_r_vs_py", value=blup_corr),
            dict(model="shrinkage", metric="grand_mean_abs_diff",
                 value=abs(grand - float(rm["grand_mean_ryoe"]))),
            dict(model="shrinkage", metric="tau_abs_diff",
                 value=abs(tau - float(rm["tau_between_rusher_sd"]))),
        ])
        parity["language"] = "parity"
        parity.to_csv(RESULTS / "metrics__shrinkage_parity__py.csv", index=False)
        gm_diff = abs(grand - float(rm["grand_mean_ryoe"]))
        print(f"[shrinkage/py] tau={tau:.3f} sigma={sigma:.3f} ICC={icc:.4f}; "
              f"vs R: BLUP corr={blup_corr:.4f}, grand-mean diff={gm_diff:.2e}")
    else:
        print(f"[shrinkage/py] tau={tau:.3f} sigma={sigma:.3f} ICC={icc:.4f} (R output not found)")


if __name__ == "__main__":
    main()
